### PR TITLE
chore: support contributing to the AWS CDK using alternative container clients

### PR DESCRIPTION
**Pull request by mrgrain**
The AWS CDK uses `docker` for a feature subset related to building assets.
For users, we already support the `CDK_DOCKER` env variable to change the container client at runtime.
However contributing to the AWS CDK currently still requires a contributor to have `docker` installed.

This PR extends the support for the `CDK_DOCKER` env variable to contributor tooling.
For example one can now run:

```console
CDK_DOCKER=finch ./scripts/foreach.sh "yarn build"
CDK_DOCKER=finch ./scripts/foreach.sh "yarn test"
```

When using `finch`, the following packages do still fail tests, due to known feature gaps:

```csv
@aws-cdk/aws-ec2
@aws-cdk/aws-lambda
@aws-cdk/aws-lambda-go
@aws-cdk/aws-lambda-python
@aws-cdk/aws-lambda-nodejs
@aws-cdk/aws-s3-deployment
```

---

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Construct Runtime Dependencies:

* [ ] This PR adds new construct runtime dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-construct-runtime-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by MrArnoldPalmer** [comment]
I believe this will resolve https://github.com/aws/aws-cdk/issues/16209, linking for posterity.

**Comment by ningziwen** [comment]
> The following packages do currently fail tests, presumingly due to some incompatibility of finch with Docker features. What exactly, will require further investigation.

Is it possible to log the finch commands when they fail? We can probably get a full list of the imcompatible gap from this.

**Comment by ningziwen** [comment]
Is the "build" in this PR mean the build of CDK binary? (as I saw you run yarn build for each in the PR description)

If we add CDK_DOCEKR to yarn build, does it mean CDK will have different binaries for each container client?

**Comment by RomainMuller** [comment]
> Is the "build" in this PR mean the build of CDK binary? (as I saw you run yarn build for each in the PR description)
> 
> If we add CDK_DOCEKR to yarn build, does it mean CDK will have different binaries for each container client?

There is no such thing as a `CDK binary`... Anyhow, this is a runtime-change... the use of `yarn build` here was to validate that the tests all still pass when using `finch`.

@mrgrain this makes me think we should fast-follow with a PR to ensure we run integration tests with both Docker & Finch (at least the container-using ones).

**Comment by mrgrain** [comment]
> I believe this will resolve #16209, linking for posterity.

@MrArnoldPalmer I don't think it does. #16209 talks about how `podman` isn't fully compatible with docker and so a few docker commands we currently use might need adjustment to be podman compatible. It's similar to how `finch` has feature gaps.

**Review by kaizencc — APPROVED** [review]
lgtm

**Comment by MrArnoldPalmer** [comment]
@mrgrain I took it as "provide a way to swap out docker for something else" whether that something else is 100% compatible with all the things we need it to do is kinda hard to control. For podman specifically we can test it, but regardless you're right lets keep it open for now.

File: scripts/check-build-prerequisites.sh [hunk 1]
```diff
@@ -74,19 +74,51 @@ else
...
+        echo "Ok"
+    else
+        die "Docker is not running"
+    fi
+
+# [finch >= 0.3.0]
+elif [ "$container_client" == "finch" ]; then
+    check_which $container_client $finch_min
+
+    # Make sure docker is running
```
Comment (ningziwen) [inline scripts/check-build-prerequisites.sh:100]: "Make sure finch is running"

**Pull request merged**